zebra-consensus/src/checkpoint.rs

```diff
@@ -324,302 +477,265 @@ impl CheckpointVerifier {
+        );
+        let range_heights: Vec<BlockHeight> = self
+            .queued
+            .range_mut(current_range)
+            .rev()
+            .map(|(key, _)| *key)
+            .collect();
+        // A list of pending valid blocks, in reverse chain order
+        let mut rev_valid_blocks = Vec::new();
+
+        // Check all the blocks, and discard all the bad blocks
+        for current_height in range_heights {
+            let valid_qblock = self.process_height(current_height, expected_hash);
+            if let Some(qblock) = valid_qblock {
+                expected_hash = qblock.block.header.previous_block_hash;
+                // Add the block to the end of the pending block list
+                // (since we're walking the chain backwards, the list is
+                // in reverse chain order)
+                rev_valid_blocks.push(qblock);
+            } else {
+                // The last block height we processed did not have any blocks
+                // with a matching hash, so chain verification has failed.
+                //
+                // TODO(teor||jlusby): log an error here?
+
+                // We kept all the matching blocks down to this height, in
+                // anticipation of the chain verifying. But the chain is
+                // incomplete, so we have to put them back in the queue.
+                //
+                // The order here shouldn't matter, but add the blocks in
+                // height order, for consistency.
+                for vblock in rev_valid_blocks.drain(..).rev() {
+                    let height = vblock
+                        .block
+                        .coinbase_height()
+                        .expect("queued blocks have a block height");
+                    self.queued.entry(height).or_default().push(vblock);
                 }
-            }
-            // Can't fail, we just checked it in the previous loop
-            expected_hash = next_parent_hash.unwrap();
-        }
 
-        // Double-check that all the block lists are empty
-        #[cfg(any(dev, test))]
-        {
-            let qrange = self.queued.range(current_range).rev();
-            for (_, qblocks) in qrange {
-                debug_assert_eq!(qblocks.len(), 0);
+                // Ensure that we're making progress
+                assert_eq!(
+                    self.previous_checkpoint_height(),
+                    old_prev_check_height,
+                    "processing must not change progress on failure"
+                );
+                let current_target = self.target_checkpoint_height();
+                assert!(
+                    current_target == WaitingForBlocks
+                        || current_target < Checkpoint(target_checkpoint_height),
+                    "processing must decrease or eliminate target on failure"
+                );
+
+                // Stop verifying, and wait for the next valid block
+                return;
```

Comment:
I don't quite follow the logic here.  We get into this arm when the range we're attempting to verify has failed, let's say `(????XYYYY]` where `Y` are matching blocks and `X` is the failing one, and then we put back all the valid blocks `YYYY` and remove `X`, so the new state is `(????_YYYY]` (a new "hole" where `X` was). What does it mean to ensure that we're making progress here?  Am I misunderstanding what this code is doing?

Reply:
You're right about what the code is doing, but the comments and assertions aren't great.

I improved them in 36999ddb50c07205fe9c9aa634da25e0ceff3280.